### ISSUE TITLE
Failed to push image: received unexpected HTTP status: 500 Internal Server Error

### ISSUE THREAD

**Issue by bparees** [openshift/origin#13955]
```
** Build Logs:
Apr 28 09:29:35.505: INFO: Running 'oc logs --config=/tmp/extended-test-docker-build-pullsecret-xtfpn-rrtrw-user.kubeconfig --namespace=extended-test-docker-build-pullsecret-xtfpn-rrtrw -f build/docker-build-1 --timestamps'
2017-04-28T13:29:02.894130000Z Pulling image centos:7 ...
2017-04-28T13:29:04.636081000Z Step 1 : FROM centos:7
2017-04-28T13:29:04.638434000Z  ---> a8493f5f50ff
2017-04-28T13:29:04.638941000Z Step 2 : ENV "OPENSHIFT_BUILD_NAME" "docker-build-1" "OPENSHIFT_BUILD_NAMESPACE" "extended-test-docker-build-pullsecret-xtfpn-rrtrw"
2017-04-28T13:29:04.699505000Z  ---> Running in 7337cde2edba
2017-04-28T13:29:04.773734000Z  ---> dae979935ad0
2017-04-28T13:29:04.778713000Z Removing intermediate container 7337cde2edba
2017-04-28T13:29:04.778998000Z Step 3 : LABEL "io.openshift.build.namespace" "extended-test-docker-build-pullsecret-xtfpn-rrtrw" "io.openshift.build.name" "docker-build-1"
2017-04-28T13:29:04.842527000Z  ---> Running in 47d20745dbdd
2017-04-28T13:29:04.902307000Z  ---> ecbdd4847ec7
2017-04-28T13:29:04.912445000Z Removing intermediate container 47d20745dbdd
2017-04-28T13:29:04.912790000Z Successfully built ecbdd4847ec7
2017-04-28T13:29:04.930863000Z Pushing image 172.30.182.180:5000/extended-test-docker-build-pullsecret-xtfpn-rrtrw/image1:latest ...
2017-04-28T13:29:07.355698000Z Pushed 0/1 layers, 3% complete
2017-04-28T13:29:27.647590000Z Pushed 1/1 layers, 100% complete
2017-04-28T13:29:31.043858000Z Registry server Address: 
2017-04-28T13:29:31.044263000Z Registry server User Name: serviceaccount
2017-04-28T13:29:31.044454000Z Registry server Email: serviceaccount@example.org
2017-04-28T13:29:31.044631000Z Registry server Password: <<non-empty>>
2017-04-28T13:29:31.044815000Z error: build error: Failed to push image: received unexpected HTTP status: 500 Internal Server Error

```

https://ci.openshift.redhat.com/jenkins/job/merge_pull_request_origin/503

**Comment by ncdc** [comment]
I'm not sure you can do much about this, other than maybe add retry logic to the registry's storage driver, or have the build try the push multiple times:

```
(*Base).PutContent" trace.id=1a830c9c-13f2-4a8a-820d-adc61f1a53cc trace.line=95 vars.name="extended-test-docker-build-pullsecret-xtfpn-rrtrw/image1" vars.uuid=5da4ea3a-c7cc-4987-ac87-e07f74b10c6d 
time="2017-04-28T13:29:30.969772273Z" level=error msg="response completed with error" err.code=unknown err.detail="gcs: storage: can't read object origin-ci-test-prtest-5a37c28-1476/registry/docker/registry/v2/repositories/extended-test-docker-build-pullsecret-xtfpn-rrtrw/image1/_uploads/5da4ea3a-c7cc-4987-ac87-e07f74b10c6d/hashstates/sha256/2851, status code: 500 Internal Server Error" err.message="unknown error"
```

**Comment by bparees** [comment]
the build actually does retry the push for certain errors already, but I guess this condition is not one of them.

**Comment by mfojtik** [comment]
@bparees do we have any smart parser for retriable errors? if we do, we should add this one. Not sure we can do something from the registry side. If you want to take this issues and fix it in build cool, otherwise i would close this.

**Comment by bparees** [comment]
I don't feel great about retrying 500 errors as a rule.  Can the registry return anything more meaningful?

**Comment by ncdc** [comment]
@bparees the last time I looked at what sort of errors the registry can return, there wasn't much flexibility, and you also had to take into account how the docker daemon processes errors from the registry and then exposes them back to the client. But I will admit that was months ago. @miminar @legionus do you have any ideas here?

**Comment by miminar** [comment]
@ncdc that's right. Upstream refused our plea for a way to forward descriptive error messages from middleware to the client.

Nevertheless, they are open to accept new generic error types associated with http codes. However, for this particular case, I don't see any better description than internal server error.

**Comment by bparees** [comment]
ok, i guess we'll retry 500 errors.  I guess worst case you've got a badly broken registry and your builds take an extra long time to fail until you fix your registry.

**Comment by csrwng** [comment]
> Not sure we can do something from the registry side.

@mfojtik are you saying that this error can't be retried on the registry side? If it's retried somewhere, I'd prefer that it be retried on the registry side, since that's a narrower set of cases. On the build end, retrying every 500 error seems heavy handed.

**Comment by malagant** [comment]
I had this issue, too today.
Just a hint with a great solution if you're having this (works also with 3.6)

https://access.redhat.com/solutions/2173071#comments

TL;DR: It's all about permissions.

**Comment by bparees** [comment]
@malagant i'm afraid that's just one of many potential causes of 500 errors from the registry, but i'm glad you were able to sort yours out!

**Comment by malagant** [comment]
@bpares You are definitely right.

**Comment by bparees** [comment]
we put in build retries, sounds like the registry can't do much, closing.

**Comment by nehagup** [comment]
I also faced the same error, in my case the registry was full. So I added archival policy to the images and made some space, it solved my problem.

**Comment by alpharameeztech** [comment]
I had the same problem, I fixed it by disabling validation.
`
kdocker run -d -p 5000:5000 --name registry -e REGISTRY_VALIDATION_DISABLED=true registry:2
`

**Comment by DRaviReddy** [comment]
for sample project its working, but for actual code its giving 500 error when i access my azure fucntion and localhost:8080 is working fine

any help would be appreciated
